server/schedule/checker/replica_checker.go

```diff
@@ -157,11 +157,13 @@ func (r *ReplicaChecker) checkMakeUpReplica(region *core.RegionInfo) *operator.O
 	}
 	log.Debug("region has fewer than max replicas", zap.Uint64("region-id", region.GetID()), zap.Int("peers", len(region.GetPeers())))
 	regionStores := r.cluster.GetRegionStores(region)
-	target := r.strategy(region).SelectStoreToAdd(regionStores)
+	target, filterByTempState := r.strategy(region).SelectStoreToAdd(regionStores)
 	if target == 0 {
 		log.Debug("no store to add replica", zap.Uint64("region-id", region.GetID()))
 		checkerCounter.WithLabelValues("replica_checker", "no-target-store").Inc()
-		r.regionWaitingList.Put(region.GetID(), nil)
+		if filterByTempState && r.regionWaitingList.Len() <= DefaultCacheSize {
```

Comment:
I think check the `list.Len` before `Put` to ensure the list not overflow without a lock is not threadsafe, And fortunally, we have already check the length in the function `regionWaitingList.Put`. 
```suggestion
		if filterByTempState  {
```

Reply:
However, I've found that here `regionWaitingList` is a LRU and it will remove the earlist item when overflow, But here we skip the latest item. So Does this matters?